### PR TITLE
Add community security audit report & contract safety improvements

### PR DESCRIPTION
**Pull request by saiboyizhan**
## Summary

Independent community security audit of `BAP578.sol` with verified code fixes. As a builder shipping a prediction market on BSC, I audited the BAP-578 reference implementation to assess integration safety and contribute back to the ecosystem.

**Audit Report**: `SECURITY-AUDIT.md` — 16 findings (2 Critical, 4 High, 3 Medium, 4 Low, 3 Informational)

**Code Fixes** (13 issues resolved in `contracts/BAP578.sol`):

| Fix | Description |
|-----|-------------|
| C-01/C-02 | `emergencyWithdraw()` restricted to unallocated funds via `totalAgentBalances` tracker |
| H-03 | Added `require(msg.value == 0)` in free mint path to prevent ETH lock |
| M-01 | Added `tokensOfOwnerPaginated()` for safe bounded queries |
| M-02 | Added `FreeMintsPerUserUpdated` event to `setFreeMintsPerUser()` |
| M-03 | Added string length validation for all metadata fields |
| L-02 | Added `fallback()` with descriptive revert |
| L-04 | Added `delete agentStates/agentMetadata` cleanup after burn |
| I-02 | Added `nonReentrant` to `fundAgent()` for defense-in-depth |
| New | Added public `burn()` — `_burn()` was previously unreachable dead code |
| New | Added `require(msg.value > 0)` to `fundAgent()` |
| New | Added `require(amount > 0)` to `withdrawFromAgent()` |
| New | Fixed `isFreeMint` fragile pattern (increment tokenId before use) |
| New | Added `uint256[39] __gap` for upgrade-safe storage |

**Design-level recommendations** (UUPS timelock, pull-based treasury, snapshot-based free mints) are documented in the audit report as suggestions for the team to evaluate.

## Test Results

All 33 tests passing (30 original + 3 new tests for emergency withdraw protection, owner withdrawal, and burn functionality).

## Files Changed

- `SECURITY-AUDIT.md` — Full audit report
- `contracts/BAP578.sol` — Bug fixes
- `test/BAP578.test.js` — Updated and new tests

### CODE REVIEW

**Comment by christelbuchanan** [comment]
Thanks @saiboyizhan we are reviewing this now. Great improvements proposed.

**Comment by christelbuchanan** [comment]
@saiboyizhan , 

This is a great PR thank you. Directionally, the audit is well-scoped and the top two criticals are real, high-impact issues for any contract that lets users deposit funds (`fundAgent()` flow). The `SECURITY-AUDIT.md` as a standalone artifact is valuable and sets a strong dev-first rubric.

We have some initial suggestions here:

**1. Please confirm `totalAgentBalances `invariants**

- Show that every balance-affecting path updates it (deposit/withdraw/burn edge cases)
- Add a test that fuzzes: multiple agents, multiple deposits/withdraws, then assert `address(this).balance >= totalAgentBalances` always holds.

**2. Burn semantics**

- Confirm `burn() `permissions (owner/approved)
- Confirm behavior when agent has balance (must revert or auto-withdraw)
- Add tests: cannot burn with balance > 0; state deleted after burn.

**3. Free mint ETH-lock fix**

- Add a test that calling free mint with `msg.value > 0` reverts with the intended message (this was a High in the audit).

**4. Pagination function**

- Add tests for `tokensOfOwnerPaginated()` boundaries (offset >= balance, limit=0, last page).

**5. Upgradeable storage gap**

- Confirm contract is actually intended to be upgradeable (UUPS is mentioned in the audit).
- Ensure the `__gap` sits at the end of storage layout and doesn’t collide with existing slots.

**Comment by saiboyizhan** [comment]
Hi @christelbuchanan, just a friendly follow-up — I've pushed the test cases covering all 5 areas you requested (commit 3224c6b), and CI is fully green (11/11 checks passing).

Let me know if there's anything else needed for this to move forward. Happy to make further adjustments if needed!

File: contracts/BAP578.sol [hunk 1]
```diff
@@ -190,9 +222,11 @@ contract BAP578 is
     /**
      * @dev Fund an agent with ETH
      */
-    function fundAgent(uint256 tokenId) external payable whenNotPaused {
+    function fundAgent(uint256 tokenId) external payable whenNotPaused nonReentrant {
         require(_exists(tokenId), "Token does not exist");
+        require(msg.value > 0, "Must send ETH");
         agentStates[tokenId].balance += msg.value;
+        totalAgentBalances += msg.value;
```
Comment (snowmanxm) [inline contracts/BAP578.sol]: I think we need to add migration balance for upgrading. Upgrading path can break withdrawals for already-funded agents

Comment (saiboyizhan) [reply contracts/BAP578.sol]: Good catch. Added `initializeV2(uint256)` using `reinitializer(2)` so deployers upgrading from V1 can set `totalAgentBalances` correctly via `upgradeToAndCall()`. See latest commit.

**Comment by saiboyizhan** [comment]
Hi @christelbuchanan @snowmanxm — friendly follow-up!

Greptile flagged a valid issue on `initializeV2` missing `onlyOwner`, which I'll push a fix for shortly (along with two minor cleanups: event ordering in `burn()` and `isFreeMint` cleanup).

Once that's pushed, all review feedback from both of you + Greptile should be fully addressed. Would love to get this across the finish line — let me know if there's anything else needed!

File: contracts/BAP578.sol [hunk 2]
```diff
@@ -216,6 +268,14 @@ contract BAP578 is
         require(success, "Withdrawal failed");
     }
 
+    /**
+     * @dev Burn an agent NFT (token owner only, balance must be 0)
+     */
+    function burn(uint256 tokenId) external onlyTokenOwner(tokenId) {
+        _burn(tokenId);
+        emit AgentBurned(tokenId, msg.sender);
+    }
```
[the remark this answers is not in the document]
Comment (snowmanxm) [reply contracts/BAP578.sol:277]: @saiboyizhan , check this

File: contracts/BAP578.sol [hunk 3]
```diff
@@ -290,12 +374,12 @@ contract BAP578 is
...
     /**
-     * @dev Emergency withdraw (owner only)
+     * @dev Emergency withdraw — only unallocated funds (not agent balances)
      */
-    function emergencyWithdraw() external onlyOwner {
-        uint256 balance = address(this).balance;
-        require(balance > 0, "No balance");
-        (bool success, ) = payable(owner()).call{ value: balance }("");
+    function emergencyWithdraw() external onlyOwner nonReentrant {
+        uint256 unallocated = address(this).balance - totalAgentBalances;
```
[the remark this answers is not in the document]
Comment (snowmanxm) [reply contracts/BAP578.sol:380]: @saiboyizhan , check this

File: contracts/BAP578.sol [hunk 4]
```diff
@@ -290,12 +374,12 @@ contract BAP578 is
...
+        require(unallocated > 0, "No unallocated balance");
+        (bool success, ) = payable(owner()).call{ value: unallocated }("");
         require(success, "Emergency withdraw failed");
     }
```
[the remark this answers is not in the document]
Comment (snowmanxm) [reply contracts/BAP578.sol:384]: @saiboyizhan , check this
